implementations/rust/kex/src/xx.rs

```diff
@@ -106,31 +111,57 @@ impl KeyExchange for SymmetricState {
         &mut self,
         secret_handle: SecretKeyContext,
         public_key: PublicKey,
-    ) -> Result<Vec<u8>, VaultFailError> {
+    ) -> Result<(SecretKeyContext, SecretKeyContext), VaultFailError> {
+        let ck = self
+            .state
+            .ck
+            .ok_or(VaultFailError::from(VaultFailErrorKind::Ecdh))?;
+
         let mut vault = self.vault.lock().unwrap();
-        vault.ec_diffie_hellman_hkdf_sha256(
+
+        let attributes_ck = SecretKeyAttributes {
+            xtype: SecretKeyType::Buffer(32),
+            purpose: SecretPurposeType::KeyAgreement,
+            persistence: SecretPersistenceType::Ephemeral,
+        };
+
+        let attributes_k = SecretKeyAttributes {
+            xtype: SecretKeyType::Aes256,
+            purpose: SecretPurposeType::KeyAgreement,
+            persistence: SecretPersistenceType::Ephemeral,
+        };
+
+        let hkdf_output = vault.ec_diffie_hellman_hkdf_sha256(
             secret_handle,
             public_key,
-            self.state.ck.as_ref(),
-            SHA256_SIZE + AES256_KEYSIZE,
-        )
+            ck,
+            vec![attributes_ck, attributes_k],
+        )?;
+
+        if hkdf_output.len() != 2 {
+            return Err(VaultFailError::from(VaultFailErrorKind::Ecdh));
+        }
+
+        Ok((hkdf_output[0], hkdf_output[1]))
     }
 
     /// mix key step in Noise protocol
-    fn mix_key<B: AsRef<[u8]>>(&mut self, hash: B) -> Result<(), VaultFailError> {
-        let hash = hash.as_ref();
-        self.state.ck.copy_from_slice(&hash[..SHA256_SIZE]);
-        let attributes = SecretKeyAttributes {
-            xtype: SecretKeyType::Aes256,
-            persistence: SecretPersistenceType::Ephemeral,
-            purpose: SecretPurposeType::KeyAgreement,
-        };
-        let key = SecretKey::Aes256(*array_ref![hash, SHA256_SIZE, AES256_KEYSIZE]);
+    fn mix_key(
+        &mut self,
+        hkdf_output: (SecretKeyContext, SecretKeyContext),
+    ) -> Result<(), VaultFailError> {
         let mut vault = self.vault.lock().unwrap();
```

Comment:
also here -- mainly concerned about catching a https://doc.rust-lang.org/std/sync/struct.PoisonError.html, though we may be able to add a `?` if the From impl exists.

Reply:
@nilslice imho unwrapping Mutex is ok, it'll only fail if another thread panicked.

Reply:
ok, let's get agreement from @mikelodder7. if it is not recoverable, then I agree. It seems that there is some material out there that suggests unwrapping it is fine: https://users.rust-lang.org/t/any-examples-of-recovering-from-a-poisoned-lock/29435/3

maybe we should go with `self.vault.lock().expect("lock failed: mutex was poisoned")` so that the user of this code is made aware, in case the other panicked thread doesn't leave much helpful detail behind.

Reply:
I personally don't like using `expect`. In my experience, it's okay to unwrap on a Mutex.